### PR TITLE
Perform value analysis on asserts on indirect loads and stores

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100550
* #100549

As per title

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1068,7 +1068,11 @@ def gen_assert_indirect_indexing(self, buffer, index, mask):
                     indirect_bounds[indirect_name[indirect_var]] = rg
 
         def gen_min_max(var, size):
-            if dynamo_config.dynamic_shapes:
+            # TODO At the moment we just give per-LoopBody bounds. This is why
+            # we may have a `var not in indirect_bounds`. We should be able
+            # to perform this analysis for a whole SchedulerNode! Same for
+            # any other analysis that uses BoundVars
+            if dynamo_config.dynamic_shapes or var not in indirect_bounds:
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:1075]: was this the case prior to the refactoring ?

Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1075]: yeah, the implementaiton just deals with one `LoopBody` at a time.

Comment (eellison) [reply torch/_inductor/codegen/triton.py:1075]: I looked into doing it across LoopBodies but all of the cases I was looking at that didn't actually help, for what it's worth

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1045,21 +1045,45 @@ def mask_loads(self, mask):
...
+
+        # Try to bound the variable via static analysis
+        # TODO We may be able to do this for dynamic shapes as well
+        if not dynamo_config.dynamic_shapes:
+            bounds = body.bounds().get_bounds()
+            indirect_bounds = {}
+            for var, rg in bounds.items():
+                var = str(var)
+                if var.startswith("set_indirect"):
+                    idx = int(var[len("set_indirect") :])
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:1066]: what does this correspond to - `idx = int(var[len("set_indirect") :])` ?

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -1073,16 +1097,24 @@ def print_min(expr):
...
                 size = print_min(sizes)
-            # The conditions need to be in parens because of Python's operator precedence.
-            # It'd be less # error-prone to use and/or/not, which is suported by triton
-            cond = f"((0 <= {var}) & ({var} < {size}))"
-            cond_print = f"0 <= {var} < {size}"
-            if not isinstance(original_index, sympy.Integer):
-                var_mask = f"({mask})" if "&" in mask else mask
-                var_mask = f" | ~{var_mask}"
+            gen_min, gen_max = gen_min_max(var, size)
+            # FooBar
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:1101]: stray comment ?

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -1045,21 +1045,45 @@ def mask_loads(self, mask):
...
+            bounds = body.bounds().get_bounds()
+            indirect_bounds = {}
+            for var, rg in bounds.items():
+                var = str(var)
+                if var.startswith("set_indirect"):
+                    idx = int(var[len("set_indirect") :])
+                    indirect_var = body.indirect_vars[idx]
+                    indirect_bounds[indirect_name[indirect_var]] = rg
+
+        def gen_min_max(var, size):
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:1070]: nit: mabye rename ? gen_check_min_and_max_bounds ?

**Review by eellison — COMMENTED** [review]
cool! have you checked how much of TIMM perf this recovers ?

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1066]: I'm just going from `set_indirectX` to the variable `indirectX`. This particular step is just trimming the `set_indirect` prefix.

File: torch/_inductor/codegen/triton.py (hunk 3 above)
Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1101]: I was thinking for a minute whether I could simplify this code as it has some repetition, but then I realized that this is its own version of the classic coding interview question "FooBar" lol

**Comment by lezcano** [comment]
I haven't checked the performance, but @ngimel said that most of the TIMM models had actual indirections (and plenty of them) so there's little we can do about that there. Also, note that this analysis (nor the `int64` -> `int32`) is performed on dynamic shapes, so there there's not much we can do.

I proposed we deactivate the `device_asserts`  when benchmarking, as that's what you will do in a production setting. I don't know what people think about this cc @ezyang @jansel @ngimel

**Comment by ezyang** [comment]
I don't think it is appropriate to disable the device asserts in prod. Prod teams will happily trade off a bit of performance to get consistent and accurate error reporting.

**Comment by lezcano** [comment]
@eellison this stack is ready for review.

**Review by eellison — APPROVED** [review]
LGTM

**Comment by lezcano** [comment]
Closing this one in favour of the more generic approach https://github.com/pytorch/pytorch/pull/100550

**Pull request closed without merging**